### PR TITLE
Add cache for core API `validators/`

### LINKED ISSUE #752: Create API Cache package
The API Cache package should cache results of certain API calls in memory, e.g. the `validators` endpoint.

It is important that this component does not create wrong results due to chain switching.
Test cases for this should be added. The component should also have a limited size for the cached elements, otherwise the node could go OOM. It should not create wrong cached results. For example the validators for the current epoch can change with every new slot commitment, so we need to invalidate the last cached result with every slot. For past epochs the values can't change, so we can keep those results in the cache.

To limit the max size, we could use https://github.com/VictoriaMetrics/fastcache.

I went through the API, for now it only seems to make sense to cache the `validators` endpoint.

### PR DESCRIPTION
**Pull request by jkrvivian**
Close #752 

* Request validators by epoch instead of slot
* Remove granularity in validators API
* Add option to set the maximum size of cache, default to 50MB
* In validators API:
   * Always request the node for validators if the `requested_Epoch == current_Epoch` and the cache result doesn't exist with key (`EpochIndex` + `SlotIndex`)
   * Keep validatorsResponse in cache `requested_Epoch < current_Epoch`
* Add unit test and docker test for validators API

### CODE REVIEW

File: pkg/requesthandler/accounts.go [hunk 1]
```diff
@@ -32,23 +35,50 @@ func (r *RequestHandler) CongestionByAccountAddress(accountAddress *iotago.Accou
...
+		registeredValidators, err = r.protocol.Engines.Main.Get().SybilProtection.OrderedRegisteredCandidateValidatorsList(epochIndex)
+	} else {
+		registeredValidators, err = r.registeredValidatorsFromCache(epochIndex)
+	}
 	if err != nil {
-		return nil, ierrors.Join(echo.ErrInternalServerError, ierrors.Wrapf(err, "failed to get ordered registered validators list for epoch %d", latestEpoch))
+		return nil, ierrors.Wrapf(err, "failed to get registered validators for epoch %d", epochIndex)
 	}
 
 	page := registeredValidators[cursorIndex:lo.Min(cursorIndex+pageSize, uint32(len(registeredValidators)))]
```
Comment (alexsporn) [inline pkg/requesthandler/accounts.go]: we need some checks here since cursorIndex can be passed from the outside

File: pkg/requesthandler/accounts.go [hunk 2]
```diff
@@ -60,7 +97,7 @@ func (r *RequestHandler) Validators(slotRange, cursorIndex, pageSize uint32) (*a
 	if int(cursorIndex+pageSize) > len(registeredValidators) {
 		resp.Cursor = ""
```
Comment (daria305) [inline pkg/requesthandler/accounts.go]: We could have single `if` here as Cursor is already empty
```
// not the last page
if int(cursorIndex+pageSize) <= len(registeredValidators) {
resp.Cursor = fmt.Sprintf("%d,%d", epochIndex, cursorIndex+pageSize)
}
```

File: pkg/requesthandler/accounts.go [hunk 3]
```diff
@@ -32,26 +35,60 @@ func (r *RequestHandler) CongestionByAccountAddress(accountAddress *iotago.Accou
...
+		registeredValidators, err = r.registeredValidatorsFromCache(epochIndex, key)
+	} else {
+		registeredValidators, err = r.registeredValidatorsFromCache(epochIndex, epochIndex.MustBytes())
+	}
+	if err != nil {
+		return nil, ierrors.Wrapf(err, "failed to get registered validators for epoch %d", epochIndex)
+	}
+
+	pageEndIndex := lo.Min(cursorIndex+pageSize, uint32(len(registeredValidators)))
+	if cursorIndex >= pageEndIndex {
```
Comment (daria305) [inline pkg/requesthandler/accounts.go]: In simpler terms this could be 
`int(cursorIndex) >= len(registeredValidators)`
and we can check is `pageSize==0` at the beginning

File: pkg/requesthandler/requesthandler.go [hunk 4]
```diff
@@ -1,26 +1,40 @@
...
+	})
 }
 
 // Shutdown shuts down the block issuer.
 func (r *RequestHandler) Shutdown() {
 	r.workerPool.Shutdown()
 	r.workerPool.ShutdownComplete.Wait()
 }
+
+func WithCacheMaxSizeOptions(size int) options.Option[RequestHandler] {
```
Comment (daria305) [inline pkg/requesthandler/requesthandler.go:36]: We should use it, or remove it

File: pkg/requesthandler/requesthandler.go [hunk 5]
```diff
@@ -1,26 +1,40 @@
...
-func New(p *protocol.Protocol) *RequestHandler {
-	return &RequestHandler{
-		workerPool: p.Workers.CreatePool("BlockHandler"),
-		protocol:   p,
-	}
+func New(p *protocol.Protocol, opts ...options.Option[RequestHandler]) *RequestHandler {
+	return options.Apply(&RequestHandler{
+		workerPool:       p.Workers.CreatePool("BlockHandler"),
+		protocol:         p,
+		optsCacheMaxSize: 2 << 20, // 2MB
```
Comment (daria305) [inline pkg/requesthandler/requesthandler.go]: What about adding it to parameters next to `restapi.ParamsRestAPI.MaxPageSize`?

File: pkg/requesthandler/accounts.go [hunk 6]
```diff
@@ -32,26 +35,60 @@ func (r *RequestHandler) CongestionByAccountAddress(accountAddress *iotago.Accou
...
-	//if !exists {
-	//	registeredValidators, err := r.protocol.Engines.Main.Get().SybilProtection.OrderedRegisteredCandidateValidatorsList(latestEpoch)
-	//	if err != nil {
-	//		return nil, ierrors.Wrapf(echo.ErrInternalServerError, "failed to get ordered registered validators list for epoch %d : %s", latestEpoch, err)
-	//	}
-	//	r.protocol.Engines.Main.Get().BlockRetainer.RetainRegisteredValidatorsCache(slotRange, registeredValidators)
-	//}
-
-	registeredValidators, err := r.protocol.Engines.Main.Get().SybilProtection.OrderedRegisteredCandidateValidatorsList(latestEpoch)
+func (r *RequestHandler) registeredValidatorsFromCache(epochIndex iotago.EpochIndex, key []byte) ([]*api.ValidatorResponse, error) {
```
Comment (daria305) [inline pkg/requesthandler/accounts.go]: It could be nicer if the cache mechanics would be more separated, like this would be the method in `cache/cache.go`
Or at least wrap the second of creating the cache in some `Store` func
```go
if registeredValidatorsBytes == nil {
// get the ordered registered validators list from engine.
registeredValidators, err := r.protocol.Engines.Main.Get().SybilProtection.OrderedRegisteredCandidateValidatorsList(epochIndex)
if err != nil {...}
err := r.registeredValidatorsCache.Store(key, registeredValidators)
```

File: pkg/requesthandler/accounts.go [hunk 7]
```diff
@@ -32,26 +35,60 @@ func (r *RequestHandler) CongestionByAccountAddress(accountAddress *iotago.Accou
...
+		registeredValidatorsBytes, err := apiForEpoch.Encode(registeredValidators, serix.WithTypeSettings(validatorResponsesTypeSettings))
+		if err != nil {
+			return nil, ierrors.Wrapf(echo.ErrInternalServerError, "failed to encode ordered registered validators list for epoch %d : %s", epochIndex, err)
+		}
+		r.registeredValidatorsCache.Set(key, registeredValidatorsBytes)
+
+		return registeredValidators, nil
+	}
+
+	validatorResp := make([]*api.ValidatorResponse, 0)
```
Comment (daria305) [inline pkg/requesthandler/accounts.go]: Following the above suggestion we could do decoding in `Cache` logic and already get `[]*api.ValidatorResponse`

**Review by daria305 — CHANGES_REQUESTED** [review]
Really nice test, if you could also remove 'Test_Validators' with its todos from coreapi_test.go after merging develop

**Review by daria305 — APPROVED** [review]
